pkg/controllers/corecluster/corecluster_controller_test.go

```diff
@@ -0,0 +1,244 @@
+			Expect(cl.Create(context.Background(), namespace)).To(Succeed())
+			testNamespaceName = namespace.Name
+
+			//TODO: implement unit tests across multiple providers
+			By("creating the testing OCP infrastructure on AWS")
+			ocpInfrastructure = configv1resourcebuilder.Infrastructure().AsAWS(testInfraName, testRegionName).Build()
+			Expect(cl.Create(ctx, ocpInfrastructure)).To(Succeed())
+
+			ocpInfrastructureCopy = ocpInfrastructure.DeepCopy()
+			ocpInfrastructureCopy.Status = configv1.InfrastructureStatus{
+				APIServerInternalURL: "https://test:8081",
+				InfrastructureName:   testInfraName,
+				Platform:             configv1.AWSPlatformType,
+				PlatformStatus: &configv1.PlatformStatus{
+					Type: configv1.AWSPlatformType,
+					AWS: &configv1.AWSPlatformStatus{
+						Region: testRegionName,
+					},
+				},
+			}
+			Expect(cl.Status().Patch(ctx, ocpInfrastructure, client.MergeFrom(ocpInfrastructureCopy))).To(Succeed())
+		})
+
+		AfterEach(func() {
+			testutils.CleanupResources(Default, ctx, testEnv.Config, cl, testNamespaceName, &configv1.Infrastructure{}, &capiv1.Cluster{}, &capav1.AWSCluster{})
+		})
+
+		Context("When there is no Core Cluster", func() {
+			BeforeEach(func() {
+				r = &CoreClusterReconciler{
+					ClusterOperatorStatusClient: operatorstatus.ClusterOperatorStatusClient{
+						Client: cl,
+					},
+					Cluster:  &capiv1.Cluster{},
+					Infra:    ocpInfrastructureCopy.DeepCopy(),
+					Platform: ocpInfrastructureCopy.Status.PlatformStatus.Type,
+				}
+			})
+
+			Context("When there is no Infra Cluster", func() {
+				It("should error saying there is no infra cluster", func() {
+					By("running the core cluster controller reconciliation")
+					_, err := r.Reconcile(ctx, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Namespace: testNamespaceName,
+							Name:      testInfraName,
+						},
+					})
+					Expect(err).To(HaveOccurred()) // TODO match error.
+
+					infraCluster := &capav1.AWSCluster{}
+					err = cl.Get(ctx, client.ObjectKey{
+						Name:      testInfraName,
+						Namespace: testNamespaceName,
+					}, infraCluster)
+					Expect(err).To(HaveOccurred()) // TODO match error.
+
+					coreCluster := &capiv1.Cluster{}
+					err = cl.Get(ctx, client.ObjectKey{
+						Name:      testInfraName,
+						Namespace: testNamespaceName,
+					}, coreCluster)
+					Expect(err).To(HaveOccurred()) // TODO match error.
+				})
+			})
+
+			Context("When there is an infra cluster", func() {
+				var infraCluster *capav1.AWSCluster
+				BeforeEach(func() {
+					infraCluster = capabuilder.AWSCluster().WithName(testInfraName).WithNamespace(testNamespaceName).Build()
+					Expect(cl.Create(ctx, infraCluster)).To(Succeed())
+				})
+
+				It("should create the core cluster", func() {
+					By("running the core cluster controller reconciliation")
+					_, err := r.Reconcile(ctx, reconcile.Request{
```

Comment:
Do we also need to have tests where the controller is running and reporting errors in another way? Rather than calling reconcile directly?

Reply:
Looking into this